### PR TITLE
convert dcp1 projects to dcp2 artifact schema (#176)

### PR DESCRIPTION
**Pull request by bmccandless**
split the migrate_dcp_one_to_dcp_two.py into two:
migrate_dcp_one_to_dcp_two.py and gather_dcp1_data.py.   
currently migrate_dcp_one_to_dcp_two.py now just handles
processing the command line arguments, then making calls

modify the migrate_dcp_one_to_dcp_two.py script to generate
json in the new Artifact Schema.  projects can be appended to
an existing json files, so it is possible to build an artifact
containing all the projects in this way.
    
added the migration_utils.py file.  This handles mapping the dcp1
data to the Artifact Schema.

added one end to end test case.  It reads a dcp1 tar.gz file, processes
it, and generates an artifact json file.  Then compares against a known
good file.

### CODE REVIEW

**Comment by mweiden** [comment]
@bmccandless great to see tests here 👍

File: dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+    def __init__(self):
+        self.name = ""
+        self.institution = ""
+
+
+class MetadataFile(MetadataBase):
+    """defines the file data, as defined in the artifact schema"""
+
+    # NOTE:  The file data is not filled in by this script.
```
Comment (maniarathi) [inline dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py]: What's the purpose of having the MetadataFile class if it is not being used anywhere?

File: dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+            return False
+
+    def add_entity(self, file_source, object_data):
+        """Add a dcp1 entity to the entity_data dict"""
+        with self.lock:
+            entity_type = get_entity_type(file_source, object_data)
+            self.entity_data.setdefault(entity_type, []).append(object_data)
+
+    def process_project(self):
+        """Process the one and only project entityyy"""
```
Comment (maniarathi) [inline dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py]: Three y's here?

File: dcp_prototype/backend/wrangling/README.md [hunk 3]
```diff
@@ -18,9 +16,16 @@ The location can be one of the following:
...
+If not specified, the output goes to standard out.
+
+If --output-file is specified, and --append is specified, then the new project
+will be combined with the existing output file.  It will add a new project if it is new,
+or replace an existing project if it is repeated.
+
 Example use:
 ```
-    python3 migrations/scripts/migrate_dcp_one_to_dcp_two.py -i <source> --threads <nthreads>
+    python3 migrations/scripts/migrate_dcp_one_to_dcp_two.py -i <source> --threads <nthreads> --output-file <output> --append
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply dcp_prototype/backend/wrangling/README.md]: It's described in the above paragraph.  --append will add a new project to an existing json file.

File: dcp_prototype/backend/wrangling/migrations/scripts/create_artifact.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        "-i",
+        "--input_directory",
+        help="An s3 prefix contains metadata files that were a part of a single DCP 1.0",
+        default="s3://hca-dcp-one-backup-data/dcp-one-copy/",
+    )
+    parser.add_argument("-o", "--output-file", help="the json output will be save at this file.", required=True)
+    arguments = parser.parse_args()
+
+    input_directory = arguments.input_directory
+    if input_directory.startswith("s3:"):
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply dcp_prototype/backend/wrangling/migrations/scripts/create_artifact.py]: Thanks, I meant to fill in the else part.   I'll do that now.

File: dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    for source_tuple, dest_attr, mapfunc in mapping:
+        source = source_data
+        try:
+            process_mapping(source, source_tuple, dest_object, dest_attr, mapfunc)
+        except KeyError as e:
+            if source_tuple not in missing:
+                missing[source_tuple] = (label, dest_attr, e)
+
+
+class DatasetMetadata(object):
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py]: Sure, I can make one file for just the Metadata classes and one for all the migration code (everything else in the file).

File: requirements-dev.txt [hunk 6]
```diff
@@ -3,3 +3,6 @@ black
 flake8-black
 chalice==1.12.0
 coverage>=2.0.15
+jsonschema
+boto3>=1.11.17
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply requirements-dev.txt]: Sure, I can alphabetize them.

File: dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py (hunk 1 above)
Comment (bmccandless) [reply dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py]: It's not used yet.  And maybe it won't be.  It's just one of the 3 objects defined in the schema.  I can remove and bring it back if we need it.

File: dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py (hunk 2 above)
Comment (bmccandless) [reply dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py]: Yeah, that's a vi fail :).   "yy" yanks the current line, but only if you'r in the command mode.

File: dcp_prototype/backend/wrangling/migrations/utils/gather_dcp1_data.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+            object_data = json.loads(buf)
+            dataset_metadata.add_entity(filename, object_data)
+            index += 1
+            if index % 1000 == 0:
+                print("processed ", index)
+
+    return dataset_metadata
+
+
+def main():
```
Comment (maniarathi) [inline dcp_prototype/backend/wrangling/migrations/utils/gather_dcp1_data.py]: This section of code seems to be exactly the same as `migrate_dcp_one_to_dcp_two.py`?

Comment (bmccandless) [reply dcp_prototype/backend/wrangling/migrations/utils/gather_dcp1_data.py]: Oh yeah,.  I had meant to remove all that.   I'll do that now.

**Pull request merged**